ykcompile/src/lib.rs

```diff
@@ -633,6 +633,86 @@ impl<TT> TraceCompiler<TT> {
     }
 
+    fn c_cast(&mut self, op: &Operand, cty: &Ty) -> Location {
+        match op {
+            Operand::Place(p) => {
+                let (loc, ty) = self.place_to_location(p, false);
+                let temp = self.create_temporary();
+                match loc {
+                    Location::Register(reg) => {
+                        match cty.size() {
+                            1 => {
+                                dynasm!(self.asm
+                                    ; mov Rb(temp), Rb(reg)
+                                );
+                            }
+                            2 => match ty.size() {
+                                1 => {
+                                    dynasm!(self.asm
+                                        ; movzx Rw(temp), Rb(reg)
+                                    );
+                                }
+                                2 | 4 | 8 => {
+                                    dynasm!(self.asm
+                                        ; mov Rw(temp), Rw(reg)
+                                    );
+                                }
+                                _ => todo!("{}", ty.size()),
+                            },
+                            4 => match ty.size() {
+                                1 => {
+                                    dynasm!(self.asm
+                                        ; movzx Rd(temp), Rb(reg)
+                                    );
+                                }
+                                2 => {
+                                    dynasm!(self.asm
+                                        ; movzx Rd(temp), Rw(reg)
+                                    );
+                                }
+                                4 | 8 => {
+                                    dynasm!(self.asm
+                                        ; mov Rd(temp), Rd(reg)
+                                    );
+                                }
+                                _ => todo!("{}", ty.size()),
+                            },
+                            8 => {
+                                match ty.size() {
+                                    1 => {
+                                        dynasm!(self.asm
+                                            ; movzx Rq(temp), Rb(reg)
+                                        );
+                                    }
+                                    2 => {
+                                        dynasm!(self.asm
+                                            ; movzx Rq(temp), Rw(reg)
+                                        );
+                                    }
+                                    4 | 8 => {
+                                        // mov reg32, reg32 in x64 automatically zero-extends to
+                                        // reg64.
+                                        dynasm!(self.asm
+                                            ; mov Rq(temp), Rq(reg)
```

Comment:
For 32bit integers you need to use `Rd` not `Rq`.

Reply:
Fixed in c89306c